core/src/main/scala/org/apache/spark/sql/rapids/tool/qualification/QualificationAppInfo.scala

```diff
@@ -494,6 +503,24 @@ class QualificationAppInfo(
       appId, hasFailures)
   }
 
+  private def getMlFuntions: Option[Seq[MLFunctions]] = {
+    stageIdToInfo.map { case (_, stageInfo) =>
+      val mlOps = checkMLOps(stageInfo.info.details)
+      if (mlOps.nonEmpty) {
+        stageInfo.mlOps = Some(mlOps)
+      }
+    }
+    val stagesWithMlOps = stageIdToInfo.filter(ml => ml._2.mlOps.nonEmpty)
+    if (stagesWithMlOps.nonEmpty) {
+      Some(stagesWithMlOps.map(
+        mlOp => MLFunctions(
+          Some(appId), mlOp._1._1, mlOp._2.mlOps.get, mlOp._2.duration.getOrElse(0))
```

Comment:
all the _1 and _2 here make this hard to read.  I don't think mlOp here is actually an op but staginfo, right?    it think this goes to my previous comment we don't need to store it in stageInfo and it would be easier to read.   Can we just have the CheckMLOps return an MLFunctions?